libgadget/fofpetaio.c

```diff
@@ -169,40 +152,96 @@ static int fof_cmp_origin(const void * c1, const void * c2) {
 }
 #endif
 
-static void fof_distribute_particles(MPI_Comm Comm) {
+static int
+order_by_type_and_grnr(const void *a, const void *b)
+{
+    const struct particle_data * pa  = (const struct particle_data *) a;
+    const struct particle_data * pb  = (const struct particle_data *) b;
+
+    if(pa->Type < pb->Type)
+        return -1;
+    if(pa->Type > pb->Type)
+        return +1;
+    if(pa->GrNr < pb->GrNr)
+        return -1;
+    if(pa->GrNr > pb->GrNr)
+        return +1;
+
+    return 0;
+}
+
+static void fof_distribute_particles(struct part_manager_type * halopartmanager, struct slots_manager_type * haloslotmanager, MPI_Comm Comm) {
     int i, ThisTask;
     MPI_Comm_rank(Comm, &ThisTask);
-    struct PartIndex * pi = mymalloc("PartIndex", sizeof(struct PartIndex) * PartManager->NumPart);
+    int NpigLocal = 0;
+    /* SlotsManager Needs initializing!*/
+    memcpy(haloslotmanager, SlotsManager, sizeof(struct slots_manager_type));
+
+    for(i = 0; i < 6; i ++) {
+        haloslotmanager->info[i].size = 0;
+        haloslotmanager->info[i].maxsize = 0;
+    }
+    /* Count how many particles we have*/
+    //#pragma omp parallel for reduction(+: NpigLocal)
+    for(i = 0; i < PartManager->NumPart; i ++) {
+        if(P[i].GrNr >= 0) {
+            NpigLocal++;
+            int type = P[i].Type;
+            /* How many of slot type?*/
+            if(type < 6 && type >= 0 && haloslotmanager->info[type].enabled)
+                haloslotmanager->info[type].maxsize++;
+        }
+    }
+
+    struct particle_data * halopart = mymalloc("HaloParticle", sizeof(struct particle_data) * NpigLocal);
+    halopartmanager->Base = halopart;
+    halopartmanager->NumPart = NpigLocal;
 
-    int64_t NpigLocal = 0;
-    int GrNrMax = -1;	/* will mark particles that are not in any group */
+    /* Note this means we will have to compact slots in the fof exchange*/
+    for(i = 0; i < 6; i ++) {
+        if(haloslotmanager->info[i].enabled) {
+            haloslotmanager->info[i].ptr = mymalloc("HaloSlots", haloslotmanager->info[i].elsize * haloslotmanager->info[i].size);
+        }
+    }
+
+    struct PartIndex * pi = mymalloc("PartIndex", sizeof(struct PartIndex) * NpigLocal);
+
+    int GrNrMax = -1;   /* will mark particles that are not in any group */
     int GrNrMaxGlobal = 0;
+    NpigLocal = 0;
+    /* Yu: found it! this shall be int64 */
+    const uint64_t taskfactor = ((uint64_t) ThisTask) * (halopartmanager->NumPart + 1Lu);
```

Comment:
Is NumPart the same on all ranks? This looks a bit dangerous. pi[.].origin shall be non-overlapping. But if oen rank has  NumPart 0 then this doesn't seem to be right.

At least need a MPI_MAX reduction for the factor multipied on ThisTask.

Reply:
taskfactor => task_origin_offset;

Reply:
Yes! A good point, I will fix it!